### PR TITLE
Ptref: Filter on headsign

### PR DESCRIPTION
**Pull request by pbougue**
Filters on vehicle journeys having given headsign.
Usage: /stop_area?headsign=CANE
Returns stop areas used by the vehicle journeys containing the headsign `CANE`

To be used from jormungandr with: filter=vehicle_journey.has_headsign(CANE)

### CODE REVIEW

File: source/ptreferential/ptreferential.cpp [hunk 1]
```diff
@@ -91,7 +91,8 @@ namespace qi = boost::spirit::qi;
         single_clause = (word >> "." >> word >> bin_op >> (word|escaped_string|bracket_string))[qi::_val = boost::phoenix::construct<Filter>(qi::_1, qi::_2, qi::_3, qi::_4)];
         having_clause = (word >> "HAVING" >> bracket_string /*'(' >> (word|escaped_string|bracket_string) >> ')'*/)[qi::_val = boost::phoenix::construct<Filter>(qi::_1, qi::_2)];
         after_clause = ("AFTER(" >> text >> ')')[qi::_val = boost::phoenix::construct<Filter>(qi::_1)];
-        filter %= (single_clause | having_clause | after_clause) % (qi::lexeme["and"] | qi::lexeme["AND"]);
+        method_clause = (word >> "." >> word >> "(" >> (word|escaped_string|bracket_string) >> ")")[qi::_val = boost::phoenix::construct<Filter>(qi::_1, qi::_2, qi::_3)];
```
Comment (TeXitoi) [inline source/ptreferential/ptreferential.cpp]: long line

File: source/ptreferential/ptreferential.cpp [hunk 2]
```diff
@@ -91,7 +91,8 @@ namespace qi = boost::spirit::qi;
...
+        filter %= (single_clause | having_clause | after_clause | method_clause) % (qi::lexeme["and"] | qi::lexeme["AND"]);
```
Comment (TeXitoi) [inline source/ptreferential/ptreferential.cpp]: long line

File: source/ptreferential/ptreferential.cpp [hunk 3]
```diff
@@ -182,6 +183,13 @@ std::vector<idx_t> get_indexes(Filter filter,  Type_e requested_type, const Data
                 }
             }
         }
+    } else if(! filter.method.empty()) {
+        if (filter.object == "vehicle_journey" && filter.method == "has_headsign") {
+            for (const VehicleJourney* vj: d.pt_data->headsign_handler.get_vj_from_headsign(filter.value)) {
+                indexes.push_back(vj->idx);
+            }
+        }
```
Comment (TeXitoi) [inline source/ptreferential/ptreferential.cpp:203]: We should error or something like that in the else, no?

File: source/ptreferential/ptreferential.h [hunk 4]
```diff
@@ -56,22 +56,24 @@ www.navitia.io
 using navitia::type::Type_e;
 namespace navitia{ namespace ptref{
 
-// Un filter est du type stop_area.uri = "kikoolol"
+// Filter is something like stop_area.uri = "kikoolol" or vehicle_journey.has_headsign("john")
```
Comment (TeXitoi) [inline source/ptreferential/ptreferential.h]: we don't have "or"

Comment (TeXitoi) [inline source/ptreferential/ptreferential.h]: rephase as we don't understand that the "or" is not in the example but between the 2 examples. And bob is stronger than john.

**Comment by stifoon** [comment]
OK for me.

File: source/ptreferential/ptreferential.cpp [hunk 5]
```diff
@@ -182,6 +187,15 @@ std::vector<idx_t> get_indexes(Filter filter,  Type_e requested_type, const Data
                 }
             }
         }
+    } else if(! filter.method.empty()) {
+        if (filter.object == "vehicle_journey" && filter.method == "has_headsign") {
+            for (const VehicleJourney* vj: d.pt_data->headsign_handler.get_vj_from_headsign(filter.value)) {
+                indexes.push_back(vj->idx);
+            }
+        } else {
+            throw parsing_error(parsing_error::partial_error,
```
Comment (xlqian) [inline source/ptreferential/ptreferential.cpp:201]: so, according to the code, only `vehicle_journey` can have method `has_headsign`, right? is that true?

Comment (TeXitoi) [inline source/ptreferential/ptreferential.cpp:201]: That's right.

File: source/ptreferential/tests/ptret_test.cpp [hunk 6]
```diff
@@ -428,3 +436,40 @@ BOOST_AUTO_TEST_CASE(find_path_test){
...
+
+    const auto res = make_query(navitia::type::Type_e::StopArea,
+                                R"(vehicle_journey.has_headsign("vehicle_journey 1"))",
+                                *(b.data));
+    BOOST_REQUIRE_EQUAL(res.size(), 2);
+    std::vector<std::string> sas;
+    for (const auto& idx: res) {
+        sas.push_back(b.data->pt_data->stop_areas.at(idx)->name);
+    }
+    std::sort(sas.begin(), sas.end());
```
Comment (xlqian) [inline source/ptreferential/tests/ptret_test.cpp]: you can use `std::set` for sas and compare it with a `std::set`, then you don't need to sort it

File: documentation/navitia.io/source/integration.rst [hunk 7]
```diff
@@ -338,6 +338,22 @@ ________
 If you specify coords in your filter, you can modify the radius used for the proximity search.
 https://api.navitia.io/v1/coverage/fr-idf/coords/2.377310;48.847002/stop_schedules?distance=500
 
+headsign
+________
+
+- Type: `String`
+
+If given, add a filter on the vehicle journeys that has the given
+value as headsign.
```
Comment (pbougue) [inline documentation/navitia.io/source/integration.rst]: I would precise : 
value as headsign (on vehicle journey itself or at a stop time).

File: source/jormungandr/tests/ptref_tests.py [hunk 8]
```diff
@@ -382,6 +382,14 @@ def test_line_filter_line_code(self):
         eq_(len(lines), 1)
         assert "1A" == lines[0]['code']
 
+    def test_line_filter_line_code_with_resource_uri(self):
+        """test filtering lines from line code 1A in the pt call with a resource uri"""
+        response = self.query('v1/coverage/main_routing_test/physical_modes/physical_mode:0x1/lines?filter=line.code=1D')
```
Comment (pbougue) [inline source/jormungandr/tests/ptref_tests.py]: line length?

File: source/jormungandr/tests/ptref_tests.py [hunk 9]
```diff
@@ -409,3 +417,24 @@ def test_route_filter_line_code(self):
...
+    def test_headsign(self):
+        """test basic usage of headsign"""
+        response = self.query('v1/coverage/main_routing_test/vehicle_journeys?headsign=vehicle_journey 0')
+        assert 'error' not in response
+        vjs = get_not_null(response, 'vehicle_journeys')
+        eq_(len(vjs), 1)
+
+    def test_headsign_with_resource_uri(self):
+        """test usage of headsign with resource uri"""
+        response = self.query('v1/coverage/main_routing_test/physical_modes/physical_mode:0x0/vehicle_journeys?headsign=vehicle_journey 0')
```
Comment (pbougue) [inline source/jormungandr/tests/ptref_tests.py]: line length?

File: source/jormungandr/tests/ptref_tests.py [hunk 10]
```diff
@@ -409,3 +417,24 @@ def test_route_filter_line_code(self):
...
+        assert 'error' not in response
+        vjs = get_not_null(response, 'vehicle_journeys')
+        eq_(len(vjs), 1)
+
+    def test_headsign_with_code_filter_and_resource_uri(self):
+        """test usage of headsign with code filter and resource uri"""
+        response = self.query('v1/coverage/main_routing_test/physical_modes/physical_mode:0x0/vehicle_journeys?headsign=vehicle_journey 0&filter=line.code=1A')
```
Comment (pbougue) [inline source/jormungandr/tests/ptref_tests.py]: line length?

**Comment by xlqian** [comment]
ok for me

**Comment by TeXitoi** [comment]
Jenkins, test this please

**Comment by pbougue** [comment]
jenkins, test this please

**Comment by TeXitoi** [comment]
jenkins, test this please

**Pull request merged**